x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/IndexLifecycleRunner.java

```diff
@@ -82,99 +82,146 @@ boolean isReadyToTransitionToThisPhase(final String policy, final IndexMetaData
 
-    private void runPolicy(IndexMetaData indexMetaData, ClusterState currentState) {
-        if (indexMetaData == null) {
-            // This index doesn't exist any more, there's nothing to execute
+    /**
+     * Run the current step that either waits for index age, or updates/waits-on cluster state.
+     * Invoked after the cluster state has been changed
+     */
+    public void runPolicyAfterStateChange(String policy, IndexMetaData indexMetaData) {
+        Settings indexSettings = indexMetaData.getSettings();
+        String index = indexMetaData.getIndex().getName();
+        LifecycleExecutionState lifecycleState = LifecycleExecutionState.fromIndexMetadata(indexMetaData);
+        if (LifecycleSettings.LIFECYCLE_SKIP_SETTING.get(indexSettings)) {
+            logger.info("skipping policy [{}] for index [{}]: {} == true", policy, index, LifecycleSettings.LIFECYCLE_SKIP);
+            return;
+        }
+        Step currentStep = getCurrentStep(stepRegistry, policy, indexMetaData, lifecycleState);
+        if (currentStep == null) {
+            logger.warn("current step [{}] for index [{}] with policy [{}] is not recognized",
+                getCurrentStepKey(lifecycleState), index, policy);
+            return;
+        }
+
+        if (currentStep instanceof TerminalPolicyStep) {
+            logger.debug("policy [{}] for index [{}] complete, skipping execution", policy, index);
+            return;
+        } else if (currentStep instanceof ErrorStep) {
+            logger.debug("policy [{}] for index [{}] on an error step, skipping execution", policy, index);
             return;
         }
-        Settings indexSettings = indexMetaData.getSettings();
-        String policy = LifecycleSettings.LIFECYCLE_NAME_SETTING.get(indexSettings);
-        runPolicy(policy, indexMetaData, currentState, false);
-    }
 
-    private void executeClusterStateSteps(Index index, String policy, Step step) {
-        assert step instanceof ClusterStateActionStep || step instanceof ClusterStateWaitStep;
-        clusterService.submitStateUpdateTask("ILM", new ExecuteStepsUpdateTask(policy, index, step, stepRegistry, nowSupplier));
+        if (currentStep instanceof PhaseCompleteStep) {
+            // Only proceed to the next step if enough time has elapsed to go into the next phase
+            if (isReadyToTransitionToThisPhase(policy, indexMetaData, currentStep.getNextStepKey().getPhase())) {
+                moveToStep(indexMetaData.getIndex(), policy, currentStep.getKey(), currentStep.getNextStepKey());
+            }
+        } else if (currentStep instanceof ClusterStateActionStep || currentStep instanceof ClusterStateWaitStep) {
+            logger.debug("[{}] running policy with current-step [{}]", indexMetaData.getIndex().getName(), currentStep.getKey());
+            clusterService.submitStateUpdateTask("ILM",
+                new ExecuteStepsUpdateTask(policy, indexMetaData.getIndex(), currentStep, stepRegistry, this, nowSupplier));
```

Comment:
I think we said we don't need ExecuteStepsUpdateTask, but rather run ClusterStateWaitStep on the incoming cluster state and if it matches move to the next step.

Reply:
Regardless of whether the `ClusterStateWaitStep` is marked as "complete" or not, we have to issue a cluster state update, so to me we might as well keep it in the midst of the cluster state update?